### ISSUE TITLE
Let user define cache location

### ISSUE THREAD

**Issue by pllim** [astropy/astropy#17507]
Original ask: Because we are not running `astropy` in someone's home directory, I want to be able to set the download/cache directory to someplace different than `~/.astropy/downloads`. How do I do that? Setting environment variables (e.g., `$HOME` or `$XDG_CONFIG_HOME`) is not an option because that may affect other apps in unintended ways.

Might be related:

* https://github.com/astropy/astropy/issues/8748
* https://github.com/astropy/astropy/issues/9557

**Comment by neutrinoceros** [comment]
Naively, I would suggest we add an option via the config file like
```
[caching]
directory = my/cache/path
``` 
(naming is tricky here because section names normally correspond to subpackages, but here it would be "config", which is both uninformative and potentially misleading)
Note that with `get_cache_dir` and `get_config_dir`'s implementations now being coupled (https://github.com/astropy/astropy/pull/17188), it'd be preferable to go with an implementation that makes sense for the configuration directory too, which can obviously not be configured via the configuration file itself, which leads me to the second option:
we could also support our own environment variables: `ASTROPY_CONFIG_DIR` and `ASTROPY_CACHE_DIR`, which would significantly reduce the chance of collision with another app. How about that, @ejeschke ?

FWIW I think #8748 and #9557 are orthogonal to this request.

**Comment by saimn** [comment]
docs already recommend XDG_CACHE_HOME: (https://docs.astropy.org/en/latest/config/index.html#getting-started)

> If for whatever reason your $HOME/.astropy directory is not accessible (i.e., you have astropy running somehow as root but you are not the root user), the best solution is to set the XDG_CONFIG_HOME and XDG_CACHE_HOME environment variables pointing to directories, and create an astropy directory inside each of those. Both the configuration and data download systems will then use those directories and never try to access the $HOME/.astropy directory.

**Comment by saimn** [comment]
and various discussions in the past:
#6803 #10244 #10738

**Comment by saimn** [comment]
but seems XDG_CACHE_HOME is actually not used...
https://github.com/astropy/astropy/pull/17188/files#r1873758135

**Comment by saimn** [comment]
> but seems XDG_CACHE_HOME is actually not used...
https://github.com/astropy/astropy/pull/17188/files#r1873758135

so bug was introduced by #16997, #17514 proposes a fix.

**Comment by ejeschke** [comment]
The problem with the XDG_CACHE_HOME solution is that it affects many other applications that use this generic mechanism.  I would propose a solution such as advocated above by @neutrinoceros where one could precisely specify the location using the astropy config machinery. In particular, to be able to do something like:
```python
from astropy import config
...
cfg = config.get_config('caching')
cfg.directory = "/my/cache/path"
```
Apologies that I didn't research the config system thoroughly to know if this is the correct type of call.  Of course, the value could default to XDG_CACHE_HOME/.astropy . But the point is that you would be able to precisely override it for cases where you want to.

**Comment by eerovaher** [comment]
Implementing the `astropy` environment variables might not be a bad idea, but it will be some time before the next feature release. In the meanwhile, it might be useful to set the XDG environment variables only for the process involving `astropy`, e.g.
```
$ XDG_CACHE_HOME=~/some_custom_dir ipython
```
It is also possible to use [set_temp_cache](https://docs.astropy.org/en/v7.0.0/api/astropy.config.set_temp_cache.html) and [set_temp_config](https://docs.astropy.org/en/v7.0.0/api/astropy.config.set_temp_config.html). They are not meant for global or permanent configuration, but using them as context managers might be good enough for many applications.

**Comment by pllim** [comment]
> it will be some time before the next feature release

6 months will arrive sooner than we realize... 😸

**Comment by nstarman** [comment]
Or 😰 .... we switch, like scipy and other major libraries, to managing downloads with [`pooch`](https://www.fatiando.org/pooch/latest/), which has very customizable cache directories and simultaneously eliminate a lot of our utils related to downloading (which IMO are rightfully in utils because they are not strictly in this project's scope and thus prime targets for replacement by well-documented, supported, and more feature-rich options).

**Comment by pllim** [comment]
Re: pooch -- That is a separate can of worms, see https://github.com/astropy/astropy-data/issues/87 for a start.

**Comment by neutrinoceros** [comment]
FWIW I have a little bit of experience with pooch (which is used in a single function in yt, `yt.load_sample`), and I never had any reason to complain about it over 4 years that it's been in use there. But it's also not a *crucial* or even *required* dependency in that context, and if I understand this discussion correctly, `pooch` would be a hard requirement in astropy. Is this right, @nstarman ?

**Comment by nstarman** [comment]
Yes. for managing the . I've also been using `pooch` for a few years now and it's excellent.

**Comment by pllim** [comment]
Please move `pooch` discussion to a new issue as that isn't the original ask here. Thanks.

**Comment by pllim** [comment]
@ejeschke , do you want to give the solution merged via https://github.com/astropy/astropy/pull/19575 a try? It should have landed on astropy nightly wheel by now. Would be nice to have your feedback before we cut v8.0.0 release so we can be sure it meets your needs. Thanks!
